quelpa.el

```diff
@@ -1839,14 +1854,25 @@ nil."
   (when (quelpa-setup-p) ;if init fails we do nothing
     (let* ((quelpa-upgrade-p (if current-prefix-arg t quelpa-upgrade-p)) ;shadow `quelpa-upgrade-p'
            (quelpa-stable-p quelpa-stable-p) ;shadow `quelpa-stable-p'
-           (cache-item (if (symbolp arg) (list arg) arg)))
+           (cache-item (if (symbolp arg) (list arg) arg))
+           version)
       (quelpa-parse-plist plist)
       (quelpa-parse-stable cache-item)
-      (quelpa-package-install arg)
+      (when (setq version (quelpa-package-install arg))
+        (setq cache-item `(,(car cache-item)
+                           ,@(plist-put (cdr cache-item) :version version))))
       (quelpa-update-cache cache-item)))
   (quelpa-shutdown)
   (run-hooks 'quelpa-after-hook))
 
+;; compatibility
+(advice-add #'package-delete :after
```

Comment:
Since it is frowned upon by some people to advice stuff directly, we could make this optional. I did this with `quelpa-use-package` by adding a function to active/deactive the advice: https://github.com/quelpa/quelpa-use-package#overriding-use-package-always-ensure
I don't mind the direct advice though. For me it's an important feature that `quelpa` can keep track of what packages the user doesn't want anymore and if it's optional the user will likely overlook it and doesn't get a consistent behavior.

Reply:
It's just generally a very bad idea to add unconditional advice to functions built-in to Emacs like this.  Even if it seems harmless, and should be so, it can cause unforeseen problems.

@kiennq You say that `:after` advice doesn't affect behavior, but that is only true as long as no bugs ever happen.  Emacs documentation shows how `:after` advice works:

    WHERE can be explained by showing the resulting new function, as the
    result of combining FUNCTION and the previous value of PLACE, which we
    call OLDFUN here:

    ‘:after’	(lambda (&rest r) (prog1 (apply OLDFUN r) (apply FUNCTION r)))

So if the advice function signals an error, it will cause an error when calling the advised function.

We want Quelpa to have a great, trouble-free reputation (I've heard people complaining about bugs, thinking it's unmaintained, etc), so I think it's important that its code "be on its best behavior" as an Emacs "citizen."  :)

Is there any other way this information could be tracked or computed, besides advising `package-delete`?

Reply:
Since user can call `package-delete` directly without `quelpa` presented, I don't think there's other  way of obtain that information.
@alphapapa, How about I wrap the `:after` advice behind an `ignore-errors` and add the function to activate/deactive advice like @steckerhalter suggested?

Reply:
> @alphapapa, How about I wrap the `:after` advice behind an `ignore-errors` 

That seems like a good idea.

> and add the function to activate/deactive advice like @steckerhalter suggested?

Do you mean a minor mode?

Reply:
In the latest iteration, I've put it as customized variable, user can customize it and have advice automatically added/removed

Reply:
Unfortunately, a `defcustom` is not the idiomatic, nor really correct, way to handle conditional advice.  Even though you used `:set`, that will not always work, e.g. many, *many* users use `setq` to set customization options, which bypasses the setters.  (Trust me, I know this from experience...and bug reports...).  This should be handled with a minor mode.

Reply:
Fine, I update it to use minor mode

Reply:
Thanks.